linklord.1
.nh
.TH LINKLORD 1 2020\-01\-08 Linux "User Manuals"
.SH NAME
.PP
\fB\fClinklord\fR \- a markdown flavored bookmark manager

.SH SYNOPSIS
.PP
.RS

.nf
linklord [\-\-settings\-dir|\-s DIR] [\-\-links\-dir|\-d DIR] [\-\-print|\-p FORMAT]|[\-\-exec|\-x FORMAT]
linklord \-\-add|\-a LINK [\-\-settings\-dir|\-s DIR] [\-\-links\-dir|\-d DIR] [\-\-category|\-c CATEGORY] [\-\-title|\-t TITLE] [\-\-add\-to\-history]
linklord [\-\-settings\-dir|\-s DIR] [\-\-links\-dir|\-d DIR] MARKDOWN\_FILE
linklord \-\-help|\-h
linklord \-\-version|\-v

.fi
.RE

.SH OPTIONS
.PP
\fB\fC\-\-settings\-dir\fR|\fB\fC\-s\fR DIR
.br
Override the environment variable:
\fBLINKLORD\_SETTINGS\_DIR\fP

.PP
\fB\fC\-\-links\-dir\fR|\fB\fC\-d\fR DIR
.br
Override the environment variable:
\fBLINKLORD\_LINKS\_DIR\fP

.PP
\fB\fC\-\-print\fR|\fB\fC\-p\fR FORMAT
.br
Print the FORMAT string to stdout when a link is
selected.  \fB\fC%u\fR and \fB\fC%t\fR in FORMAT will be
replaced with URL and TITLE of the selected link.

.PP
\fB\fC\-\-exec\fR|\fB\fC\-x\fR FORMAT
.br
the FORMAT string will get evaluated when a link
is selected.  \fB\fC%u\fR and \fB\fC%t\fR in FORMAT will be
replaced with URL and TITLE of the selected link.

.PP
\fB\fC\-\-add\fR|\fB\fC\-a\fR LINK
.br
Add URL to the \fIdatabase\fP\&.

.PP
\fB\fC\-\-category\fR|\fB\fC\-c\fR CATEGORY
.br
If set the prompt for category when using the
\fB\fC\-\-add\fR option will get bypassed. The link will
get saved to CATEGORY, (the relative path to a
file in LINKLORD\_DIR).

.PP
\fB\fC\-\-title\fR|\fB\fC\-t\fR TITLE
.br
If set the prompt for title when using the
\fB\fC\-\-add\fR option will get bypassed.  The value of
TITLE will be used as title for the link.

.PP
\fB\fC\-\-add\-to\-history\fR
.br
If set links will get added to the history file
when the \fB\fC\-\-add\fR option is used.

.PP
\fB\fC\-\-help\fR|\fB\fC\-h\fR
.br
Show help and exit.

.PP
\fB\fC\-\-version\fR|\fB\fC\-v\fR
.br
Show version and exit.

.SH environment variables
.SS LINKLORD\_SETTINGS\_DIR
.PP
Path to where linklord store its configuration
and log files
(\fB\fCsettings\fR,\fB\fCactions\fR,\fB\fC\&.history\-\&l\fR,\fB\fC\&.history\-\&a\fR,\fB\fChistory\-c\fR,\fB\fC\&.log\fR).
Defaults to \fB\fCXDG\_CONFIG\_HOME/.config/linklord\fR,
and can be overridden with the command line option
\fB\fC\-\-settings\-dir|\-s\fR\&.

.SS LINKLORD\_LINKS\_DIR
.PP
Where the links are stored, defaults to:
\fB\fCLINKLORD\_SETTINGS\_DIR/links\fR, and can be
overridden with the command line option
\fB\fC\-\-links\-dir|\-d\fR\&.

.SH files
.PP
If \fBLINKLORD\_SETTINGS\_DIR\fP doesn't exist when
\fB\fClinklord\fR is executed, the directory will get
created with a \fB\fCsettings\fR and \fB\fCaction\fR file with
the default values set for all global variables.

.SH global variables
.PP
Every time \fB\fClinklord\fR is executed the
\fB\fCLINKLORD\_SETTINGS\_DIR/settings\fR will get sourced
and any variables declared in the file will be
used as global variables in \fB\fClinklord\fR\&.

.PP
\fBLINKLORD\_SETTINGS\fP

.PP
.RS

.nf
#!/bin/bash

\_history\_size=5
\_spliton="linklord was here"
\_char\_blacklist="][<'"
\_prefixlink=" " 
\_prefixfile=" " 
\_prefixfolder=" "
\_menu\_browse=(dmenu \-p "select link: ")
\_menu\_action=(dmenu \-p "select action: ")
\_menu\_add\_title=(dmenu \-p "title for url: ")
\_menu\_add\_category=(dmenu \-p "store in category: ")

# shellcheck disable=SC2034

.fi
.RE

.SS \fB\fC\_history\_size\fR
.PP
How many selected links to remember, these will
be listed before other links, files and
directories. If value is \fB\fC0\fR, no limit will be
used.

.SS \fB\fC\_spliton\fR
.PP
String used to indicate where \fB\fClinklord\fR has
appended links to a markdown document.

.SS \fB\fC\_char\_blacklist\fR
.PP
The characters in this string is not allowed in
titles.

.SS \fB\fC\_prefix*\fR
.PP
Prefix for the different types used in the
browser menu. By default unicode characters from
the iconfont "Font Awesome" is used, but changing
these settings makes it easy to customize.

.SS \fB\fC\_menu\_*\fR
.PP
The commands used to spawn the different menus.
Must be entered as arrays (within parenthesis).
Any menu program that accepts a list as input can
be used. Tested with \fB\fCfzf\fR (terminal only),
\fB\fCrofi\fR, \fB\fCi3menu\fR and \fB\fCdmenu\fR\&.

.SH the "database"
.PP
\fB\fClinklord\fR searches for links in all files that
are not prefixed with a \fB\fC\&.\fR in
\fBLINKLORD\_LINKS\_DIR\fP\&. The links are stored in
markdown format like this:

.PP
\fBLINKLORD\_DIR\fP/file1

.PP
.RS

.nf
[title1]: URL\-1
[title2]: URL\-A
[title3]: URL\-1

.fi
.RE

.PP
\fBLINKLORD\_DIR\fP/subdir/file2

.PP
.RS

.nf
[title4]: URL\-1
[title2]: URL\-A
[title5]: URL\-C

.fi
.RE

.PP
The title is linked to the URL, but not vice
verse. Multiple records can have the same URL but
different titles(\fBtitle1\fP,\fBtitle3\fP and
\fBtitle4\fP). Records where both the URL and the
title is the same can occur in multiple files
(\fBtitle2\fP). Think of the files as
\fBcategories\fP\&. When a markdown file is searched
it only matches the titles and as soon as it finds
a match it insert the corresponding URL, this is
why one title can't match multiple URLs.

.PP
By using the command line option \fB\fC\-\-add URL\fR with
\fB\fClinklord\fR a verification of the title will be
performed before adding the link. If \fB\fC\-\-title
TITLE\fR and \fB\fC\-\-category CATEGORY\fR is not set, a
menu will prompt for that info.

.PP
\fB\fClinklord \-\-add URL\-D\fR would result in first a
prompt for the title:
.br
> title for URL\-D:

.PP
lets say we enter "title1". Now a prompt for
\fBcategory\fP will get displayed together with a
list of all \fBcategories\fP (\fIi.e files in
LINKLORD\_LINKS\_DIR\fP). We can select one of the
entries in the list or enter the name for a new
category. if we assume the file \fB\fCbudlabs\fR exist,
and we select that. We would first get an error
message since that already exist with a different
URL. A new prompt to re\-enter the title, will be
shown lets enter "MyTitle".  The category will be
the same so no need to enter that twice. This
title is valid and the file
\fB\fCLINKLORD\_LINKS\_DIR/budlabs\fR will now look like
this:

.PP
.RS

.nf
[github]: https://github.com/budlabs
[youtube]: https://youtube.com/c/dubbeltumme
[MyTitle]: URL\-D

.fi
.RE

.SH the "browser"
.PP
If neither a markdown file or the \fB\fC\-\-add\fR option
is used when \fB\fClinklord\fR is invoked it will instead
display a menu with the links in
\fBLINKLORD\_LINKS\_DIR\fP,  it will also list all
categories (files).
.br
budlabs github
\[la]https://github.com/budlabs\[ra] If the following three files
exist in \fBLINKLORD\_LINKS\_DIR\fP:

.PP
.RS

.nf
\- `LINKLORD\_DIR/`
    \- `subdir/`
        \- `file2`  
          [title4]: URL\-1
          [title2]: URL\-A
          [title5]: URL\-C

    \- `file1`
      [title1]: URL\-1
      [title2]: URL\-A
      [title3]: URL\-1

    \- 'budlabs'
      [github]: https://github.com/budlabs
      [youtube]: https://youtube.com/c/dubbeltumme
      [MyTitle]: URL\-D

.fi
.RE

.PP
the list would look something like this:

.PP
.RS

.nf
L github
L youtube
L MyTitle
L title1
L title2
L title3
F budlabs
F file1
D subdir


.fi
.RE

.PP
(\fB\fCL\fR == link, \fB\fCF\fR == file, \fB\fCD\fR == directory)

.PP
notice that no links from \fBsubdir/file2\fP is
included and that the directory name (\fBsubdir\fP)
is. If a file is selected, a new list with only
the links in that file will get listed. If a
directory is selected all files, links and
directories within the selected directory will get
listed.

.SS actions
.PP
If a link is selected a new menu with actions
will get displayed. Actions are defined in the
file \fB\fCLINKLORD\_SETTINGS\_DIR/actions\fR .  Or with
the commandline options \fB\fC\-\-print FORMAT\fR or
\fB\fC\-\-exec FORMAT\fR\&. When the commandline options are
used the action menu will not be displayed. Each
action consists of two parts: The action and a
FORMAT:

.PP
\fB\fCLINKLORD\_SETTINGS\_DIR/actions\fR

.PP
.RS

.nf
print %t \- %u
exec browser %u

.fi
.RE

.RS
.IP \(bu 2
\fBprint\fP: prints FORMAT to \fB\fCstdout\fR
.IP \(bu 2
\fBexec\fP: evaluates FORMAT.

.RE

.PP
FORMAT has two special symbols that will get
expanded when the action is executed:
* \fB%u\fP \- expands to the selected links URL

.RS
.IP \(bu 2
\fB%t\fP \- expands to the selected links title
.br


.RE

.PP
After the action is executed, the selected link
will also get added to the history
(\fB\fCLINKLORD\_SETTINGS\_DIR/.history\-f\fR), the links in
the history will get added to the top of the list
next time \fB\fClinklord\fR is executed for browsing.

.PP
When a markdown file (*a file with either \fBmd\fP
or \fBmarkdown\fP*) is passed as an argument to
\fB\fClinklord\fR, that markdown file will get searched
for declared but not referenced markdown links. If
those references is found in the file based
database they will get appended to the markdown
file.

.PP
\fBLINKLORD\_LINKS\_DIR\fP/budlabs

.PP
.RS

.nf
[github]: https://github.com/budlabs
[youtube]: https://youtube.com/c/dubbeltumme

.fi
.RE

.PP
\fBarticle.md\fP

.PP
.RS

.nf
# example direct reference
This first link will get ignored, 
it already has direct url reference: [link1](https://example.com)

# missing reference
But this one: [MyTitle] doesn't so linklord will search for an url
matching "MyTitle".

# not missing reference
[link5] is already referenced in the file, 
it will be ignored by linklord.

[link5]: https://example.com

# referencing reference
[link3][youtube] "link3" has the reference "youtube", 
so it will get ignored, but youtube it self lacks a 
reference so that will get searched for.

.fi
.RE

.PP
The command: \fB\fClinklord article.md\fR would print
the following message:
.br
>     NO URL: MyTitle
.br
>     ADDED: youtube

.PP
\fBarticle.md\fP will get updated to this:

.PP
\fBarticle.md\fP

.PP
.RS

.nf
# example direct reference
This first link will get ignored, 
it already has direct url reference: [link1](https://example.com)

# missing reference
But this one: [MyTitle] doesn't so linklord will search for an url
matching "MyTitle".

# not missing reference
[link5] is already referenced in the file, 
it will be ignored by linklord.

[link5]: https://example.com

# referencing reference
[link3][youtube] "link3" has the reference "youtube", 
so it will get ignored, but youtube it self lacks a 
reference so that will get searched for.

[linklord was here]: #
[youtube]: https://youtube.com/c/dubbeltumme

.fi
.RE

.PP
The reference \fB\fC[linklord was here]\fR is inserted
for indicating what has been automatically
appended, everything below this line will get
overwritten by \fBlinklord\fP every time this
document is processed. The string \fBlinklord was
here\fP, can be changed by setting the \fB\fC\_spliton\fR
variable in \fB\fCLINKLORD\_SETTINGS\_DIR/settings\fR

.SH ENVIRONMENT
.PP
\fB\fCXDG\_CONFIG\_HOME\fR

.PP
defaults to: $HOME/.config

.PP
\fB\fCLINKLORD\_SETTINGS\_DIR\fR

.PP
defaults to: $XDG\_CONFIG\_HOME/linklord

.PP
\fB\fCLINKLORD\_LINKS\_DIR\fR

.PP
defaults to: $LINKLORD\_SETTINGS\_DIR/links

.SH DEPENDENCIES
.PP
\fB\fCbash\fR \fB\fCgawk\fR \fB\fCsed\fR

.PP
budRich https://github.com/budlabs/linklord
\[la]https://github.com/budlabs/linklord\[ra]

.SH SEE ALSO
.PP
dmenu(1), fzf(1), rofi(1),


linklord.bash
#!/usr/bin/env bash

___printversion(){
  
cat << 'EOB' >&2
linklord - version: 2020.01.07.5
updated: 2020-01-07 by budRich
EOB
}


# environment variables
: "${XDG_CONFIG_HOME:=$HOME/.config}"
: "${LINKLORD_SETTINGS_DIR:=$XDG_CONFIG_HOME/linklord}"
: "${LINKLORD_LINKS_DIR:=$LINKLORD_SETTINGS_DIR/links}"


main(){


  _menu_browse=(dmenu -p "select link: ")
  _menu_action=(dmenu -p "select action: ")
  _menu_add_title=(dmenu -p "title for url: ")
  _menu_add_category=(dmenu -p "store in category: ")
  _find_options=(-maxdepth 1 -mindepth 1 -not -name ".*")

  [[ -n "${__o[settings-dir]}" ]] \
    && LINKLORD_SETTINGS_DIR="${__o[settings-dir]}"

  [[ -n "${__o[links-dir]}" ]] \
    && LINKLORD_LINKS_DIR="${__o[links-dir]}"
  
  local sd=$LINKLORD_SETTINGS_DIR
  local ld=$LINKLORD_LINKS_DIR

  [[ -d $sd ]] || createconf "$sd"
  [[ -f $sd/settings ]] && . "$sd/settings"

  _history_links="$sd/.history-l"
  _history_actions="$sd/.history-a"
  _history_categories="$sd/.history-c"
  _reportfile="$sd/.log"
  _actionfile="$sd/actions"

  : "${_history_size:=10}"
  : "${_spliton:="linklord was here"}"
  : "${_char_blacklist:="][<>"}"
  : "${_prefixlink:=" " }"
  : "${_prefixfile:=" " }"
  : "${_prefixfolder:=" "}"

  if [[ -f $* ]]; then
    appendlinks "$*"
  elif [[ -n ${__o[add]} ]]; then
    addlink "${__o[add]}"
  elif [[ -z $* ]]; then
    [[ -d $ld ]] || ERX "$ld doesn't exist"
    listlinks "$ld"
  fi

}




___printhelp(){
  
cat << 'EOB' >&2
linklord - a markdown flavored bookmark manager


SYNOPSIS
--------
linklord [--settings-dir|-s DIR] [--links-dir|-d DIR] [--print|-p FORMAT]|[--exec|-x FORMAT]
linklord --add|-a LINK [--settings-dir|-s DIR] [--links-dir|-d DIR] [--category|-c CATEGORY] [--title|-t TITLE] [--add-to-history]
linklord [--settings-dir|-s DIR] [--links-dir|-d DIR] MARKDOWN_FILE
linklord --help|-h
linklord --version|-v

OPTIONS
-------

--settings-dir|-s DIR  
Override the environment variable:
LINKLORD_SETTINGS_DIR


--links-dir|-d DIR  
Override the environment variable:
LINKLORD_LINKS_DIR


--print|-p FORMAT  
Print the FORMAT string to stdout when a link is
selected.  %u and %t in FORMAT will be replaced
with URL and TITLE of the selected link.


--exec|-x FORMAT  
the FORMAT string will get evaluated when a link
is selected.  %u and %t in FORMAT will be replaced
with URL and TITLE of the selected link.


--add|-a LINK  
Add URL to the database.


--category|-c CATEGORY  
If set the prompt for category when using the
--add option will get bypassed. The link will get
saved to CATEGORY, (the relative path to a file in
LINKLORD_DIR).


--title|-t TITLE  
If set the prompt for title when using the --add
option will get bypassed.  The value of TITLE will
be used as title for the link.


--add-to-history  
If set links will get added to the history file
when the --add option is used.



--help|-h  
Show help and exit.


--version|-v  
Show version and exit.

EOB
}


addlink() {
  local trg str dunstifyis
  local url="$1"
  local title="${__o[title]:-}"
  local category="${__o[category]:-}"

  : "${title:=$(echo -n | "${_menu_add_title[@]}")}"

  [[ -z $title ]] || : "${category:=$(
    {
      allfiles
      [[ -f "$_history_categories" ]] && cat "$_history_categories"
      printf '%s\n' "${_files[@]}" | sed "s;${LINKLORD_LINKS_DIR}/;;g"
    } | awk '!a[$0]++' | "${_menu_add_category[@]}"
           
  )}"

  # close Error message from previous add link if
  # dunstifyis (installed)
  [[ ${dunstifyis:=$(command -v dunstify)} ]] \
    && dunstify -C 123454

  [[ -z $category || -z $title ]] && ERX add link canceled

  addtohistory "$category" "$_history_categories"
  # addtohistory
  trg="$LINKLORD_LINKS_DIR/$category"

  assert="$(verifytitle "$title" "$url" "$trg")"
  if [[ $assert = "$trg" ]]; then
    ERM "$title already exist in $category"

  elif [[ -n $assert ]]; then

    if [[ $dunstifyis ]]; then
      dunstify -r 123454 -u critical "$assert"
    else
      notify-send -u critical "$assert"
    fi

    addlink "$url" --category "$category"
    exit
  else

    mkdir -p "${trg%/*}"

    str="$(printf '[%s]: %s' "$title" "$url")"

    echo "$str" >> "$trg"

    ((__o["add-to-history"] == 1)) \
      && addtohistory "$str" "$_history_links" "$_history_size"

    ERM "$(printf 'added url: %s\nas %s in %s\n' \
                  "$url" "$title" "$category"
          )"
  fi
}

addtohistory() {
  local name="$1" history="$2" limit="${3:-666}" tmpf

  if [[ -f $history ]]; then

    tmpf="$(mktemp)"

    awk -v limit="${limit}" -v name="$name" '
      BEGIN {a[name]=1 ; print name}
      !a[$0]++ && limit > ++i {print}
    ' "$history" > "$tmpf"

    mv -f "$tmpf" "$history"
  else
    echo "$name" > "$history"
  fi
}

# store path to all files in array '_files'
# if it doesn't already exist
# separet with null char -d ''
allfiles() {
  [[ -z ${_files[0]} ]] && {
    declare -ga _files
    
    readarray -t -d '' _files  \
      < <(find "$LINKLORD_LINKS_DIR" \
               -mindepth 1     \
               -not -name ".*" \
               -type f         \
               -print0         \
         )
  }        
}
appendlinks() {

  local trgpath="$*" trgfile tdir regex linklist

  [[ ! $trgpath =~ ^/ ]] && trgpath="$PWD/$trgpath"
  trgfile="${trgpath##*/}"

  rm -f "$_reportfile"
  linklist="$(getlinks "$trgpath")" 

  [[ -f "$_reportfile" ]] && ERM "$(sort -u "$_reportfile")"
  
  [[ -n $linklist ]] && tdir="$(mktemp -d)" && (
    cd "$tdir" || ERX could not create temp dir
    cp "$trgpath" .

    regex="$(printf '/[[]%s[]]: #/' "$_spliton")"
    csplit "$trgfile" "$regex" > /dev/null 2>&1 \
      && mv -f xx00 "$trgfile"

    {
     echo "$(< "$trgfile")"
     printf '\n[%s]: #\n%s' "$_spliton" "$linklist"
    } > "$trgpath"
    
    rm -rf "$tdir"  
  )
}

createconf() {
local trgdir="$1"
declare -a aconfdirs

aconfdirs=(
"$trgdir/links"
)

mkdir -p "$1" "${aconfdirs[@]}"

cat << 'EOCONF' > "$trgdir/actions"
print %t - %u
EOCONF

cat << 'EOCONF' > "$trgdir/settings"

_history_size=5
_spliton="linklord was here"
_char_blacklist="[]<'"
_prefixlink=" " 
_prefixfile=" " 
_prefixfolder=" "
_menu_browse=(dmenu -p "select link: ")
_menu_action=(dmenu -p "select action: ")
_menu_add_title=(dmenu -p "title for url: ")
_menu_add_category=(dmenu -p "store in category: ")

# shellcheck disable=SC2034
EOCONF

cat << 'EOCONF' > "$trgdir/links/budlabs"
[linklord wiki]: https://github.com/budlabs/linklord/wiki
[budlabs github]: https://github.com/budlabs
[budlabs youtube channel]: https://youtube.com/c/dubbeltumme
EOCONF

}
set -E
trap '[ "$?" -ne 77 ] || exit 77' ERR

ERM(){

  local mode

  getopts xr mode
  case "$mode" in
    x ) urg=critical ; prefix='[ERROR]: '   ;;
    r ) urg=low      ; prefix='[WARNING]: ' ;;
    * ) urg=normal   ; mode=m ;;
  esac
  shift $((OPTIND-1))

  msg="${prefix}$*"

  if [[ -t 2 ]]; then
    echo "$msg" >&2
  else
    notify-send -u "$urg" "$msg"
  fi

  [[ $mode = x ]] && exit 77
}

ERX() { ERM -x "$*" ;}
ERR() { ERM -r "$*" ;}
ERH(){
  {
    ___printhelp 
    [[ -n "$*" ]] && printf '\n%s\n' "$*"
  } >&2 
  exit 77
}
getlinks() {

  local targetfile="$1"
  
  allfiles

  awk -v reportfile="$_reportfile" -v targetfile="$targetfile" -F "[][]" '

    # first line not yaml start
    FNR == 1 && FILENAME == targetfile && $0 !~ /^---$/ {dashcount=2}
    /^---$/     {dashcount++}
    /^```|~~~/ {codeblock++}
    
    NF == 3 && dashcount < 1 {links[$2]=$0}

    /./ && dashcount > 1 && codeblock%2 == 0  {

      # ignore inline code
      gsub(/`[^`]+`/,"")

      # ignore image links
      gsub(/![[][^]]+[]]]/,"")
      for (i=2;i<NF;i+=2) {

        # dont include referenced
        if ($(i+1) ~ /^:/)
          delete found[$i]

        # dont include direct url []() or ref []:
        # dont bother with links not in the database
        else if (($(i+1) ~ /^\s*[^(]/ || !$(i-1)) ) {
          # [][] - $(i+1) is nothing
          # NF-1 != i to not include block at end of line
          # ![] - image, previous block ends with !
          if (!(!$(i+1) && (NF-1) != i) && $(i-1) !~ /!$/)
            name = gensub(/^.*[/]/,"",1,FILENAME)
            if ($i in links) {
              printf("%s: %s: %s\n",name, "ADDED", $i) > reportfile
              found[$i]=links[$i]
            }  else {
              printf("%s: %s: %s\n",name, "NO URL", $i) > reportfile
            }
        }
      }
    }
    END { for (k in found) {print found[k]} }
  ' "${_files[@]}" "$1"
}

geturl() {

  local title="$1"

  allfiles

  awk -F "[][]" -v title="$title" '
    $2 == title {
      print(gensub(/^.*[]]:\s*/,"",1,$0))
      exit
    }
  ' "${_files[@]}"
}

linkaction() {
  local url title printformat choice hstr

  title="${1#$_prefixlink}"
  url="$(geturl "$title")"

  printformat="${__o[exec]:-${__o[print]:-}}"

  [[ -z $printformat ]] && {

    [[ -f $_actionfile ]] \
      || { echo "print %u" > "$_actionfile" ;}

    choice="$(
      {
        [[ -f "$_history_actions" ]] && cat "$_history_actions"
        cat "$_actionfile"
      } | awk '!a[$0]++' | "${_menu_action[@]}"
    )"

    [[ -z $choice ]] && ERX no action selected

    addtohistory "$choice" "$_history_actions"
    __o["${choice%% *}"]=1
    printformat="${choice#* }"

  }

  hstr="$(printf '[%s]: %s' "$title" "$url")"
  addtohistory "$hstr" "$_history_links" "$_history_size"

  printformat="${printformat//%u/$url}"
  printformat="${printformat//%t/$title}"

  if  [[ -n ${__o[exec]} ]]; then
    eval "${printformat}"
  elif [[ -n ${__o[print]} ]]; then
    echo "$printformat"
  fi
}

listlinks() {
  local cur="$1" trg choice
  local format="${_prefixlink}%s\n"
  # local format="<span lang='%s'></span>%s\n"

  choice="$(
  {
    {
      # list all links ('[TITLE]: URL')
      [[ $cur = "$LINKLORD_LINKS_DIR" ]] && [[ -f $_history_links ]] \
        && cat "$_history_links"
      
      if [[ -f $cur ]]; then
        cat "$cur"
      else
        find "$cur" "${_find_options[@]}" -type f -exec cat '{}' \;
      fi
      # format linklist and remove duplicates with awk
    } | awk -v format="$format" -F "[][]" '
          NF == 3 && !a[$2]++ {
            printf(format, $2)
          }
        '

    # list all files and directories if $cur is a dir
    [[ -d $cur ]] && {
      find "$cur" "${_find_options[@]}" -type f -printf "${_prefixfile}%f\n"
      find "$cur" "${_find_options[@]}" -type d -printf "${_prefixfolder}%f\n"
    }

  } | "${_menu_browse[@]}"
  )"

  [[ -z $choice ]] && ERX nothing selected

  # test if choice has markup, if true, its a link
  if [[ $choice =~ ^${_prefixlink} ]];then
   linkaction "$choice"
  else
    # trim prefix, append trg to cur to get path
    trg="${choice#${_prefixfile}}" trg="${trg#${_prefixfolder}}"
    trg="$cur/$trg"

    # if path is neither a file or directory
    [[ -d $trg || -f $trg ]] || ERX i dont know what "${trg}" is
    
    listlinks "$trg"
  fi
}

verifytitle() {
  local title="$1" url="$2" file="$3"

  # test non accepted characters
  [[ -n $_char_blacklist && $title =~ ([${_char_blacklist}]) ]] && {
    echo "'${BASH_REMATCH[1]}' in $title, illegal character "
    return 1
  }

  allfiles
  # same title can be entered in any number of files
  # but only once in each file. A title can only be used for one url
  awk  -v title="$title" -v url="$url" -v file="$file" -F "[][]" '
    $2 == title {
      currenturl   = gensub("^." $2 ".:[[:space:]]*","",1,$0)
      currenttitle = $2
      msg=""
      
      if (currenturl != url) {
        msg = sprintf("%s is used for: %s in %s",
                      title, currenturl, FILENAME)
      } else if (FILENAME == file) {
        msg = FILENAME
      }

      if (msg != "") {
        print msg
        exit
      }
    }
  ' "${_files[@]}"
}

declare -A __o
options="$(
  getopt --name "[ERROR]:linklord" \
    --options "s:d:p:x:a:c:t:hv" \
    --longoptions "settings-dir:,links-dir:,print:,exec:,add:,category:,title:,add-to-history,help,version," \
    -- "$@" || exit 77
)"

eval set -- "$options"
unset options

while true; do
  case "$1" in
    --settings-dir | -s ) __o[settings-dir]="${2:-}" ; shift ;;
    --links-dir  | -d ) __o[links-dir]="${2:-}" ; shift ;;
    --print      | -p ) __o[print]="${2:-}" ; shift ;;
    --exec       | -x ) __o[exec]="${2:-}" ; shift ;;
    --add        | -a ) __o[add]="${2:-}" ; shift ;;
    --category   | -c ) __o[category]="${2:-}" ; shift ;;
    --title      | -t ) __o[title]="${2:-}" ; shift ;;
    --add-to-history ) __o[add-to-history]=1 ;; 
    --help       | -h ) ___printhelp && exit ;;
    --version    | -v ) ___printversion && exit ;;
    -- ) shift ; break ;;
    *  ) break ;;
  esac
  shift
done

[[ ${__lastarg:="${!#:-}"} =~ ^--$|${0}$ ]] \
  && __lastarg="" 


main "${@:-}"


